### ISSUE TITLE
API packaging

### ISSUE THREAD

**Issue by sacdallago** [PokemonGoers/PokeData#21]
If you want to test your API, I suggest the following scheme:

We have a server where we can host your API, but you won't have access to it (LRZ is very strict on this). What we can do, though, is the following: if you package your API into a [Docker](https://www.docker.com/) instance (through a .Dockerfile that reads the config.json from some given path), I will set up a hook to listen to changes on the `develop` branch and respwan docker workers as code is merged into develop (more: https://docs.docker.com/docker-hub/builds/ ).

This way, every time you approve a PR, you can check the new running code.

**Comment by swathi-ssunder** [comment]
I must say I am not familiar with Docker and have just started reading about it. 
For the above suggested task, I did setup Docker on my machine and tested our current application on it using the DockerFile attached here. 
[Dockerfile.zip](https://github.com/PokemonGoers/PokeData/files/421871/Dockerfile.zip)

Let me know if this is on the right track.

Also regarding reading from the config.json file, does that refer to reading the credentials(eg: for the twitter api) or something else. 
Excuse me if this is something very basic but I am trying to figure it myself too.

**Comment by sacdallago** [comment]
1. This skills are going beyond :) This is industry standard. If you can do this and will apply for a job at some stage in the near future, you will be in the top 20%. Also it is handy in day-to-day development.
2. You are on the right track! Take your time to understand how to do it, don't need to do it today, just soon!
3. Remember to specify what it means to run `npm start` (last line of your .dockerfile, you define that in `package.json` in the scripts section).
4. Yes! You will ideally have a `config.json` in the same folder where the `.dockerfile` will be with all the confidential data that shouldn't be displayed online (twitter keys, database and what not) and this file will be built with the docker build. For docker to auto-build the project though, you would need to pass it some mock config.json (for example giving a read-only database user, no twitter keys, aso.).
   When I spawn the instance though, I should be able to change the database information and twitter to something meaningful again, and docker allows this by passing, for example, environment variables: https://docs.docker.com/engine/reference/commandline/run/ See the `--env-file value` flag.
5. If you are wondering "how the hell can I use environment variables in a config file" --> https://github.com/sacdallago/pssh/blob/master/configuration.js.template#L9  **Important**: the file is a `.js` file, not a simple `.json` ;)

**Comment by swathi-ssunder** [comment]
Thanks for the pointers. I shall proceed in this direction and will maybe get back if I have any questions after I understand it better :)

**Comment by swathi-ssunder** [comment]
Should I commit the Dockerfile to git or does it have to be sent via email(along with the env-file)?
Refer #32.

**Comment by sacdallago** [comment]
The .dockerfie should not contain any sensible info, right?
In that case just push it to the repo with a PR.

The env file on the other hand should contain only sensible info so please send that to me via mail :) I'll try this all out later today

**Comment by swathi-ssunder** [comment]
Done. Refer #35 . 
I have also mailed you the env file.

**Comment by swathi-ssunder** [comment]
I have sent a mail to christian.dallago@tum.de. Is this the right one?

**Comment by sacdallago** [comment]
Yes @swathi-ssunder, sorry for not answering yet but I haven't had time to try it out on the production server and set up the whole pipeline.

I might just brute-force try one of these days :)

**Comment by swathi-ssunder** [comment]
@sacdallago - No problem. I just wanted to ensure that it is delivered :)

**Comment by sacdallago** [comment]
Ok.
- I set up an automatic build on docker. Every time you merge code into `develop` or `master` a build process will be spawned. You can see all of this on https://hub.docker.com/r/pokemongoers/pokedata/tags/
  1. `develop` will build a later version of `pokemongoers/pokedata:develop`
  2. `master` will build `pokemongoers/pokedata:latest` or simply `pokemongoers/pokedata`

Ideally you should be able to do:

``` bash
   docker run --env-file /PATH/TO/ENV/FILE -p 49160:8080 --name apidev -d pokemongoers/pokedata:develop npm start
```

   Notice: `npm start` is the command that will be executed when the machine is run! If in a later point in time I were to run `docker start apidev` it would execute again `npm start` using the env file etc.

Another example I tried was: `docker run --env-file /home/cdallago/pokemon/env -p 45364:8080 --name apilisten -d pokemongoers/pokedata:develop npm run listen`. It "worked", but by `docker logs apilisten` I get:

``` bash
$ docker logs apilisten
npm info it worked if it ends with ok
npm info using npm@2.14.2
npm info using node@v4.0.0
npm info prelisten pokemon-api@0.0.1
npm info listen pokemon-api@0.0.1

> pokemon-api@0.0.1 listen /usr/src/app
> node scripts/listen.js

INFO rarePokemon
module.js:338
    throw err;
    ^

Error: Cannot find module 'async'
    at Function.Module._resolveFilename (module.js:336:15)
    at Function.Module._load (module.js:286:25)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (/usr/src/app/app/controllers/filler/rarePokemon.js:6:13)
    at Module._compile (module.js:434:26)
    at Object.Module._extensions..js (module.js:452:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (/usr/src/app/scripts/listen.js:19:18)
    at Module._compile (module.js:434:26)
    at Object.Module._extensions..js (module.js:452:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)

npm info pokemon-api@0.0.1 Failed to exec listen script
npm ERR! Linux 4.2.0-41-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "listen"
npm ERR! node v4.0.0
npm ERR! npm  v2.14.2
npm ERR! code ELIFECYCLE
npm ERR! pokemon-api@0.0.1 listen: `node scripts/listen.js`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the pokemon-api@0.0.1 listen script 'node scripts/listen.js'.
npm ERR! This is most likely a problem with the pokemon-api package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node scripts/listen.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls pokemon-api
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /usr/src/app/npm-debug.log
```

Try to look into that, please
- The result of running `npm start` is available at http://test.got.show (for now a `Cannot GET /` ?)

**Comment by sacdallago** [comment]
Hint: the execution of the command via `npm run listen` might not happen in the `workingdir`. I'm **NOT SURE** about that, but it's the first thing that pops to my mind.

Also: I'm missing still data in the ENV file, like database credentials for production (which is probably why I cannot get / ?)

**Comment by swathi-ssunder** [comment]
1. Error: Cannot find module 'async'
   This error was because the async package was not included in the package.json. Maybe someone installed it locally without using a --save, so the package.json was not updated. 
   I have corrected this in the package.json now.
2. Storing to the shared MLAB database was not implemented at the time I sent the env file and hence those credentials are missing in that file. I have now modified the config file to accept these credentials from the env file.

So once my pull request #57 is merged, I will test in the new build.

**Comment by swathi-ssunder** [comment]
Have found some more issues with this, I am still working on them.

**Comment by sacdallago** [comment]
@swathi-ssunder did you try out the build by pulling from hub.docker.com? Aka not by running `docker build` but directly by `docker run ...`?
Isn't it super cool?!?!?! :joy: sorry! But I'm always super excited about advocating docker! Once you realize how useful it is :stuck_out_tongue_closed_eyes: 

btw, my script (this is not how I'm gonna do it in the end, but it was useful to try things out quickly) to do things is:

``` bash
#!/bin/bash

# Stop all containers
docker stop $(docker ps -a -q)
# Delete all containers
docker rm $(docker ps -a -q)
# Delete all images
docker rmi $(docker images -q)

docker run --env-file /home/cdallago/pokemon/env -p 49160:8080 --name apidev -d pokemongoers/pokedata:develop

exit 0
```

Unfortunately there's no way to autopull the latest build without removing the previous images from the disk https://github.com/docker/docker/issues/13331

**Comment by sacdallago** [comment]
FYI.: The latest build exectues and terminates at the end (stopping the container).

``` bash
$docker start -i apidev
npm info it worked if it ends with ok
npm info using npm@2.14.2
npm info using node@v4.0.0
npm info prestart pokemon-api@0.0.1
npm info start pokemon-api@0.0.1

> pokemon-api@0.0.1 start /usr/src/app
> NODE_ENV=production node app.js

npm info poststart pokemon-api@0.0.1
npm info ok
```

**Comment by vivek-sethia** [comment]
Thanks for sharing the log. Since the database credentials are missing, mongoose gets disconnected and hence the container terminates. Form what I understood is that we have to use environment file for passing such information.  This can be done by this command  
`docker run --env-file /PATH/TO/ENV/FILE -p 49160:8080 --name apidev -d pokemongoers/pokedata:develop npm start`.

**Comment by swathi-ssunder** [comment]
@sacdallago - Using the below mentioned steps, I was able to get it running docker on my machine.
1. First I build the docker file locally using 
      docker build -t <username>/<appname> .
2. After that , docker run command was executed
   docker run --env-file /PATH/TO/ENV/FILE -p 49160:8080 --name apidev -d <username>/<appname>
3. docker start -i apidev
   and the port started listening (the container was not stopped)
4. Used http://0.0.0.0:49160/api/pokemons to fetch the pokemon details and got a JSON response.

However, when I did the same after pulling from hub.docker.com(excluding the build step), the 4th step did not work. I mean, I was not able to access the url. 

So I am trying to figure out where it is actually listening.

**Comment by sacdallago** [comment]
That's weird.
Make sure the build on docker is the same as the one that you are using now (aka you are building from develop branch in git) and that when you pull from docker you pull the develop tag (`:develop` in a previous comment).

Run my script if necessary, so you delete all previous running containers and avoid name clashes

**Comment by swathi-ssunder** [comment]
I am also not sure what is going on either.
I am using the exact same build as on docker(from develop).
And I am indeed deleting all previous containers. So it is equivalent to what is there in your script too.
The container shows as running. The only issue is that I am unable to access the URL from eg: Postman or Rest Client.

**Comment by vivek-sethia** [comment]
@sacdallago  This might be a silly question. But even I tried doing the same thing, but I am not able to understand which url is to be used for calling the api which is hosted on the server using the docker. Since in the case of local, I was able to call it by `localhost:8080/api/pokemons` after `npm start`

**Comment by sacdallago** [comment]
Nope. It is a different port!

```
docker run --env-file /home/cdallago/pokemon/env -p 49160:8080 --name apidev -d pokemongoers/pokedata:develop
```

Means: Port 8080 inside the container maps to port 49160 of the local machine.  
Are you using osx by any chance? Because in Linux the above means `localhost:49160` = `container:8080`, but in osx the docker-machine gets some IP in the form `192..` meaning you can access it via `192..:49160`
To find out the address https://docs.docker.com/machine/reference/ip/ the machine that you use should be called `default` aka `docker-machine ip default`

**Comment by sacdallago** [comment]
I set up an automatic deployment on my own docker cloud ( http://pokedata.c4e3f8c7.svc.dockerapp.io:65014 ). It should get redeployed every time it is re-build.
Maybe in a few days I get to set it up so that it can be accessible from test.got.show, am waiting that that server get's green lights and then I need to set up the dedicated account on docker cloud.

**Comment by swathi-ssunder** [comment]
Sure.:)
However this is currently responding with "Cannot GET /".
Is there anything to be done from our end?

**Comment by sacdallago** [comment]
http://pokedata.c4e3f8c7.svc.dockerapp.io:65014/api/pokemon

**Comment by sacdallago** [comment]
You should definetly do something about the default route! ( / ) :D
Maybe put the apidocs documentation on the root!

**Comment by swathi-ssunder** [comment]
Ah. Sorry. I just tried with the old api.. /api/pokemons. I did not realise it is updated with the latest build.
Thanks :)

**Comment by swathi-ssunder** [comment]
Sure, I shall do that.

**Comment by sacdallago** [comment]
That's the whole point of having this pipeline :dancer: and not me having to redeploy the thing manually every time, because let's face it.. I'm super lazy :joy: :joy: :joy: :joy: :joy: :joy: :joy: :joy:

**Comment by swathi-ssunder** [comment]
@sacdallago - Maybe we can close this issue now ? :)

**Comment by vivek-sethia** [comment]
@sacdallago We still have to run our listen scripts on Docker so data can be stored continuously from different sources like twitter, pokerdar, pokesniper, skiplagged, pokecrew.

**Comment by sacdallago** [comment]
@vivek-sethia yup that is no big issue, I was just hoping the chair's node would be up by now.. I'll try to set up some temporary thing tomorrow, although I hate temporary

**Comment by sacdallago** [comment]
@vivek-sethia  I deployed the twitter listener but none of the others, because: which other listeners can I implement without having to set up the proxy story?
I should be able to do this by tomorrow (the ones without proxy) and then the one with the proxy we will see as soon as I can set up the proxies...

**Comment by vivek-sethia** [comment]
@sacdallago As of now, `rarePokemon` listener can also be started as it is not dependent on proxy.

**Comment by jonas-he** [comment]
@sacdallago all other ones can also be deployed (no need to adapt anything), it will just be slower without proxy. Also as the Machine Learners want more data (see #71 ) this would be a good idea.

**Comment by sacdallago** [comment]
I will do this then.. Can someone update me on database growth and stuff? If we hit the 500mb limit on mlab shit hits the fan
